cmd/specdiff/differ.go

```diff
@@ -0,0 +1,283 @@
+		panic("searchArrayAndSliceType called with invalid type")
+	}
+
+	for i := 0; i < (arrayNode.Len()); i++ {
+		childNode := arrayNode.Index(i)
+		if childNode.IsZero() {
+			continue
+		}
+		if childNode.Kind() == reflect.String {
+			changePath.fieldPath.push(childNode.String())
+			addToDiffProto(diffProto, changePath)
+			changePath.fieldPath.pop()
+			continue
+		}
+		if endpoint, ok := childNode.Interface().(diff.Endpoint); ok {
+			changePath.fieldPath.push(handleEndpointStruct(endpoint))
+			addToDiffProto(diffProto, changePath)
+			changePath.fieldPath.pop()
+			continue
+		}
+		return fmt.Errorf("array child node %v is not supported", childNode)
+	}
+	return nil
+}
+
+func searchStructType(structNode reflect.Value, diffProto *rpc.Diff, changePath *change) error {
+	if structNode.Kind() != reflect.Struct {
+		panic("searchStructType called with invalid type")
+	}
+
+	if vd, ok := structNode.Interface().(diff.ValueDiff); ok {
+		handleValueDiffStruct(vd, diffProto, changePath)
+		return nil
+	}
+	for i := 0; i < structNode.NumField(); i++ {
+		tag, ok := structNode.Type().Field(i).Tag.Lookup("json")
+		if !ok {
+			// Fields that don't have a JSON name aren't part of the diff.
+			continue
+		}
+
+		// Empty fields in the diff are redundant. Skip them.
+		childNode := structNode.Field(i)
+		if childNode.IsZero() {
+			continue
+		}
+
+		// Struct field tags have the format "fieldname,omitempty". We only want the field name.
+		fieldName := strings.Split(tag, ",")[0]
+
+		err := handleStructField(childNode, fieldName, diffProto, changePath)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func handleStructField(value reflect.Value, name string, diffProto *rpc.Diff, changePath *change) error {
+	// Empty fields in the diff are redundant. Skip them.
+	if value.IsZero() {
+		return nil
+	}
+
+	switch name {
+	case "added", "deleted", "modified":
+		changePath.changeType = name
+	default:
+		changePath.fieldPath.push(name)
+		defer changePath.fieldPath.pop()
+	}
+
+	err := searchNode(value, diffProto, changePath)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func handleEndpointStruct(ed diff.Endpoint) string {
+	Method := getAtomicType(reflect.ValueOf(ed.Method))
+	Path := getAtomicType(reflect.ValueOf(ed.Path))
+	return fmt.Sprintf("{method.%s path.%s}", Method, Path)
+}
+
+func handleValueDiffStruct(vd diff.ValueDiff, diffProto *rpc.Diff, changePath *change) {
+	fromValue := getAtomicType(reflect.ValueOf(vd.From))
+	toValue := getAtomicType(reflect.ValueOf(vd.To))
+
+	diffProto.Modification[changePath.fieldPath.String()] = &rpc.Diff_ValueChange{
+		From: fromValue,
+		To:   toValue,
+	}
+}
+
+func isAtomicType(node reflect.Value) bool {
+	switch node.Kind() {
+	case reflect.Float64:
+		return true
+	case reflect.String:
+		return true
+	case reflect.Bool:
+		return true
+	case reflect.Int:
+		return true
+	default:
+		return false
+	}
+}
+
+func getAtomicType(node reflect.Value) string {
```

Comment:
Until now I couldn't find the right words for these, but I think these types are "scalar types" and types like maps/slices/arrays are "composite" or "compound" types.

I think this function can have a more descriptive name. Something like `stringifyScalar(scalar reflect.Value) string`, `scalarString(scalar reflect.Value) string`, or similar. It should match the naming we use for stringifying the endpoint struct.